xds/internal/xdsclient/client_new.go

```diff
@@ -76,59 +84,73 @@ func newWithConfig(config *bootstrap.Config, watchExpiryTimeout time.Duration, i
 // NewWithConfigForTesting returns an xDS client for the specified bootstrap
 // config, separate from the global singleton.
 //
+// The second return value represents a close function which the caller is
+// expected to invoke once they are done using the client.  It is safe for the
+// caller to invoke this close function multiple times.
+//
 // # Testing Only
 //
 // This function should ONLY be used for testing purposes.
-func NewWithConfigForTesting(config *bootstrap.Config, watchExpiryTimeout, authorityIdleTimeout time.Duration) (XDSClient, error) {
+// TODO(easwars): Document the new close func.
+func NewWithConfigForTesting(config *bootstrap.Config, watchExpiryTimeout, authorityIdleTimeout time.Duration) (XDSClient, func(), error) {
 	cl, err := newWithConfig(config, watchExpiryTimeout, authorityIdleTimeout)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
-	return &clientRefCounted{clientImpl: cl, refCount: 1}, nil
+	return cl, grpcsync.OnceFunc(cl.close), nil
 }
 
 // NewWithBootstrapContentsForTesting returns an xDS client for this config,
 // separate from the global singleton.
 //
+// The second return value represents a close function which the caller is
+// expected to invoke once they are done using the client.  It is safe for the
+// caller to invoke this close function multiple times.
+//
 // # Testing Only
 //
 // This function should ONLY be used for testing purposes.
-func NewWithBootstrapContentsForTesting(contents []byte) (XDSClient, error) {
+func NewWithBootstrapContentsForTesting(contents []byte) (XDSClient, func(), error) {
 	// Normalize the contents
 	buf := bytes.Buffer{}
 	err := json.Indent(&buf, contents, "", "")
 	if err != nil {
-		return nil, fmt.Errorf("xds: error normalizing JSON: %v", err)
+		return nil, nil, fmt.Errorf("xds: error normalizing JSON: %v", err)
 	}
 	contents = bytes.TrimSpace(buf.Bytes())
 
+	// closeFunc returned to callers decrements the reference count on the
+	// returned client, and when it reaches zero, removed it from the map.
+	closeFunc := func(c *clientRefCounted, key string) func() {
+		return grpcsync.OnceFunc(func() {
```

Comment:
This seems to be more complicated than it needs to be?  Can you instead refactor the code so the two non-error returns are just a single return instead, and do this there, but without the closure that returns a closure?

Reply:
Hmm... Not sure if this is what you meant. But it looks like more repetitive code than before.

Reply:
No, I meant  to combine the two happy paths so the closure only needs to be written once as well.  Something like this:


```go
	clientsMu.Lock()
	defer clientsMu.Unlock()

	c := clients[string(contents)]
	if c != nil {
		c.incrRef()
	} else {

		// Maybe pull this into its own function
		bcfg, err := bootstrap.NewConfigFromContentsForTesting(contents)
		if err != nil {
			return nil, nil, fmt.Errorf("xds: error with bootstrap config: %v", err)
		}

		cImpl, err := newWithConfig(bcfg, defaultWatchExpiryTimeout, defaultIdleAuthorityDeleteTimeout)
		if err != nil {
			return nil, nil, err
		}
		c = &clientRefCounted{clientImpl: cImpl, refCount: 1}
		// (to here)

		clients[string(contents)] = c
	}

	return c, grpcsync.OnceFunc(func() {
		clientsMu.Lock()
		defer clientsMu.Unlock()
		if c.decrRef() == 0 {
			c.close()
			delete(clients, string(contents))
		}
	}), nil
```

Or like this perhaps:

```go
func getOrMakeClient(...) (*clientRefCounted, error) {
	clientsMu.Lock()
	defer clientsMu.Unlock()

	if c := clients[string(contents)]; c != nil {
		return c, nil
	}
	bcfg, err := bootstrap.NewConfigFromContentsForTesting(contents)
	if err != nil {
		return nil, fmt.Errorf("xds: error with bootstrap config: %v", err)
	}

	cImpl, err := newWithConfig(bcfg, defaultWatchExpiryTimeout, defaultIdleAuthorityDeleteTimeout)
	if err != nil {
		return nil, err
	}
	c := &clientRefCounted{clientImpl: cImpl, refCount: 0}
	clients[string(contents)] = c
	return c, nil
}


....
	c, err := getOrMakeClient(...)
	// forward err
	c.incrRef()
	return c, grpcsync.OnceFunc(func() {
		clientsMu.Lock()
		defer clientsMu.Unlock()
		if c.decrRef() == 0 {
			c.close()
			delete(clients, string(contents))
		}
	}), nil
```